File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findMaxAverage = function(nums, k) {
    let maxAvg = average = i = 0;
    
    while(i < k) {
        average += nums[i];
        i++;
    }    
    average /= k;
    maxAvg = average;
    
    while(i < nums.length) {
        average = (average * k - nums[i - k] + nums[i])/k;
        maxAvg = Math.max(maxAvg, average);
        i++;
    }
    
    return maxAvg;
};

File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.c
double findMaxAverage(int* nums, int numsSize, int k) {
//     set a left, right pointers, currSum and ans 
    int left = 0, right = 0, currSum = 0;
    double ans = -INFINITY;
    
//     start right iteration till the end of the array
    for(right; right < numsSize; right++) {
        currSum += nums[right];
        
//     while (right - left + 1 < k increase left)
        while(right - left + 1 > k) {
            currSum -= nums[left];
            left++;
        }
        
//     find the average and compare with ans and store the maximum
        if(right - left + 1 == k) {
            double currAvg = (double)currSum / k;
            ans = ans > currAvg ? ans : currAvg;
        }
    }
    
//     return ans
    return ans;
}